### DISCUSSION TITLE
how to correctly create a dict subclass with typing

### DISCUSSION THREAD

**Discussion by zzzeek** [python/typing#1033] [Q&A] [answered]
we have a lot of classes that subclass `dict` for the purpose of storing keys and values in different ways, providing different ways of returning defaults, etc.

I'm not able to create a custom subclass of `typing.Dict` where all methods are accepted by mypy.

Currently, trying to figure out the `.values()` method.  Mypy insists that the return type here be the `dict_values` type.  However this is a private Python type; while it seems to be somewhat part of "builtins", it's not actually there and you can't import it without typing and runtime errors.

What's the correct return type for `dict.values()`, or is it no longer good practice to subclass Dict and I need to instead create a wrapping object with a real dict on the inside (this would be very unfortunate and surprising?)  ?     Or am I supposed to build a custom protocol that matches dict_values somehow?   typing.ValuesView is not accepted if you are subclassing `Dict`.

example below

```py
# Module "builtins" has no attribute "dict_values"
from builtins import dict_values  # type: ignore

import time
from typing import Any
from typing import cast
from typing import Dict, List
from typing import Tuple
from typing import TypeVar
from typing import ValuesView


_KT = TypeVar("_KT", bound=Any)
_VT = TypeVar("_VT", bound=Any)



class MyDict(Dict[_KT, _VT]):
    """a dictionary that stores values as a two-tuple with a timestamp.

    the exposed values however are those that were originally passed in.

    """

    def __getitem__(self, __k: _KT) -> _VT:
        super_dict = cast(Dict[_KT, Tuple[_VT, float]], super())
        try:
            value = super_dict.__getitem__(__k)
        except KeyError:
            raise
        else:
            return value[0]

    def __setitem__(self, __k: _KT, v: _VT) -> None:
        super_dict = cast(Dict[_KT, Tuple[_VT, float]], super())
        return super_dict.__setitem__(__k, (v, time.time()))

    # values like a normal person - return a list
    # Return type "List[_VT]" of "values" incompatible with return type "dict_values[_KT, _VT]" in supertype "dict"
    # Return type "List[_VT]" of "values" incompatible with return type "ValuesView[_VT]" in supertype "Mapping"

    # def values(self) -> List[_VT]:
    #     super_dict = cast(Dict[_KT, Tuple[_VT, float]], super())
    #     return [v[0] for k, v in super_dict.items()]

    # try to use real ValuesView
    # Return type "ValuesView[_VT]" of "values" incompatible with return type "dict_values[_KT, _VT]" in supertype "dict"

    # def values(self) -> ValuesView[_VT]:
    #     super_dict = cast(Dict[_KT, Tuple[_VT, float]], super())
    #     return ValuesView({k: v[0] for k, v in super_dict.items()})

    # try to use the dict_values mypy is claiming -
    # not possible, because we cannot import private dict_values
    def values(self) -> dict_values[_KT, _VT]:
        super_dict = cast(Dict[_KT, Tuple[_VT, float]], super())
        return ValuesView({k: v[0] for k, v in super_dict.items()})

md: MyDict[str, str] = MyDict()

# fails, revealed type is Any, because we had to ignore type
# on dict_values
reveal_type(md.values())
```

**Comment by AlexWaygood** [comment, marked as answer]
Yeah this was broken in 3.10, see here: https://bugs.python.org/issue46399

We're trying to figure out some solutions!

**Reply by zzzeek** [reply]
oh wow, OK.  I googled like mad to find this issue, but i didnt think it was that high up the chain.  OK thanks!

**Reply by AlexWaygood** [reply]
No worries! It's a very suboptimal situation at the moment, and a tricky problem to solve, but  hopefully things will improve soon :)

**Reply by AlexWaygood** [reply]
FYI, I think my proposed solution (to expose `dict_keys` in the `types` module) is pretty close to being rejected by the core devs. If you have any thoughts — either for or against my suggested solution! — I'd encourage you to chime in on the BPO issue 🙂 (in a friendly and constructive way — obviously please don't start a pile-on!)

**Reply by zzzeek** [reply]
the first place I'd look would be how does pylance handle this?  or do they have the same bug.  seemed like they didn't.

**Reply by zzzeek** [reply]
also, doesn't dict_keys / dict_values implement the interface given by typing.KeysView / typing.ValuesView (I assume they are protocols)?  why wouldn't the typeshed stubs use KeysView/ ValuesView for how these methods are typed and not expose private implementation details of Python?

**Reply by AlexWaygood** [reply]
Pylance also uses typeshed stubs, so if there isn't the same issue with pylance, it's *probably* because they're special-casing it somehow. But I don't really know -- @erictraut might know more?

**Reply by zzzeek** [reply]
OK pyright doesn't report an error because it appears to lack the level of checking that is testing the overridden method against the return type of the superclass.  I can make the method like this:

```
    def values(self) -> str:
        return "hi"
```

and pyright says OK it returns "str", it doesn't care about the fact that Dict.values() returns dict_values.

I still think linking key_values / dict_values to typing.KeysView / typing.ValuesView is the most correct approach.

**Reply by erictraut** [reply]
@zzzeek, pyright will flag this as an error if you enable the `reportIncompatibleMethodOverride` check (or enable strict type checking mode, which automatically includes that check).

**Comment by zzzeek** [comment]
i know this is a different question but similar area, what's the correct typing for setdefault()?   if I have a dict[str, str], how is the None value implied by setdefault accommodated?   

```py
import typing
from typing import Dict

x: Dict[str, str] = {"a": "b"}

# we would think typing would raise an error for this
q = x.setdefault("c")

if typing.TYPE_CHECKING:
    # Revealed type is "builtins.str*"
    reveal_type(q)

# which is wrong.  it's None
assert q is not None
```

**Reply by JelleZijlstra** [reply]
I think https://github.com/python/typeshed/blob/master/stdlib/typing.pyi#L480 is buggy. It needs an overload for the case where the second argument is not specified.
